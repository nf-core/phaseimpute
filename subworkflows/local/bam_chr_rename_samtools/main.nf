include { SAMTOOLS_REHEADER           } from '../../../modules/nf-core/samtools/reheader'
include { SAMTOOLS_INDEX              } from '../../../modules/nf-core/samtools/index'

workflow BAM_CHR_RENAME_SAMTOOLS {
    take:
    ch_bam          // channel: [ [id], bam, index, prefix ]

    main:

    ch_versions = Channel.empty()

    // Rename the chromosome with or without prefix
    SAMTOOLS_REHEADER(
        ch_bam.map{
            meta, bam, index, prefix ->
            if (prefix == "nochr") {
                cmd = 'sed -E "s/^(@SQ.*\\tSN:)chr/\\1/"'
            } else if (prefix == "chr") {
                cmd = 'sed -E "s/^(@SQ.*\\tSN:)([0-9]+|X|Y|MT|M)/\\1chr\\2/"'
            } else {
                error "Invalid chr_prefix: ${prefix}"
            }
            [meta, bam, index, cmd]
        }, // channel: [ [id], bam, index, cmd]
    )
    ch_versions = ch_versions.mix(SAMTOOLS_REHEADER.out.versions.first())

    SAMTOOLS_INDEX(SAMTOOLS_REHEADER.out.bam)
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())

    ch_bam_renamed = SAMTOOLS_REHEADER.out.bam
        .combine(SAMTOOLS_INDEX.out.bai, by:0)

    emit:
    bam_renamed    = ch_bam_renamed        // [ [id], bam, csi ]
    versions       = ch_versions           // channel: [ versions.yml ]
}
