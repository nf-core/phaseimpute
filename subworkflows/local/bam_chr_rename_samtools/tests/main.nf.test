nextflow_workflow {

    name "Test Subworkflow BAM_CHR_RENAME_SAMTOOLS"
    script "../main.nf"

    config "./nextflow.config"

    workflow "BAM_CHR_RENAME_SAMTOOLS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/bam_chr_rename_samtools"
    tag "bam_chr_rename_samtools"

    tag "samtools"
    tag "samtools/reheader"
    tag "samtools/index"
    tag "gawk"

    test("Should remove chr prefix") {
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "test_paired"],
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true)
                    ],
                    [
                        [id: "test_MT"],
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam_bai'], checkIfExist:true)
                    ],
                ])
                input[1] = Channel.of("nochr")
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("Should add chr prefix") {
        setup {
            run("BAM_CHR_RENAME_SAMTOOLS", alias: "PREPROCESS") {
                script "../main.nf"
                process {
                    """
                    input[0] = Channel.fromList([
                        [
                            [id: "test_paired"],
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true)
                        ],
                        [
                            [id: "test_MT"],
                            file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam'], checkIfExist:true),
                            file(params.test_data['homo_sapiens']['illumina']['test_illumina_mt_bam_bai'], checkIfExist:true)
                        ],
                    ])
                    input[1] = Channel.of("nochr")
                    """
                }
            }
        }

        when {
            workflow {
                """
                input[0] = PREPROCESS.out.bam_renamed
                input[1] = Channel.of("chr")
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
