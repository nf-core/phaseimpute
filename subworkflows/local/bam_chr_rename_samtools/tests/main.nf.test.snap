{
    "Should remove chr prefix": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "test_MT_withchr.bam:md5,33f16cad4bdac734717255f29d98e9b6",
                        "test_MT_withchr.bam.bai:md5,dbd8f1a5df58115b6f64a9312133d79d"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "test_paired_withchr.bam:md5,c6908df90282cc95e846eee31476bb47",
                        "test_paired_withchr.bam.bai:md5,77524c609a3001cdca9c7d10dcc14b6a"
                    ]
                ],
                "1": [
                    "versions.yml:md5,42dd37f7dc7e85848f3bb65119dae337",
                    "versions.yml:md5,f91d09d8f0579548146084c6bdb1df5e"
                ],
                "bam_renamed": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "test_MT_withchr.bam:md5,33f16cad4bdac734717255f29d98e9b6",
                        "test_MT_withchr.bam.bai:md5,dbd8f1a5df58115b6f64a9312133d79d"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "test_paired_withchr.bam:md5,c6908df90282cc95e846eee31476bb47",
                        "test_paired_withchr.bam.bai:md5,77524c609a3001cdca9c7d10dcc14b6a"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,42dd37f7dc7e85848f3bb65119dae337",
                    "versions.yml:md5,f91d09d8f0579548146084c6bdb1df5e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-06-18T15:37:07.410027428"
    },
    "Should add chr prefix": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "test_MT_withchr.bam:md5,438766aaee32ee392f000d3d89569a0d",
                        "test_MT_withchr.bam.bai:md5,e94a4e26b532fba2901189f325405e45"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "test_paired_withchr.bam:md5,ed6c8a1c7f8e79e321e61057aa736f5b",
                        "test_paired_withchr.bam.bai:md5,71049dbf4968601dce88352ff34de528"
                    ]
                ],
                "1": [
                    "versions.yml:md5,42dd37f7dc7e85848f3bb65119dae337",
                    "versions.yml:md5,f91d09d8f0579548146084c6bdb1df5e"
                ],
                "bam_renamed": [
                    [
                        {
                            "id": "test_MT"
                        },
                        "test_MT_withchr.bam:md5,438766aaee32ee392f000d3d89569a0d",
                        "test_MT_withchr.bam.bai:md5,e94a4e26b532fba2901189f325405e45"
                    ],
                    [
                        {
                            "id": "test_paired"
                        },
                        "test_paired_withchr.bam:md5,ed6c8a1c7f8e79e321e61057aa736f5b",
                        "test_paired_withchr.bam.bai:md5,71049dbf4968601dce88352ff34de528"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,42dd37f7dc7e85848f3bb65119dae337",
                    "versions.yml:md5,f91d09d8f0579548146084c6bdb1df5e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-06-18T15:37:19.417850864"
    }
}